### PR TITLE
core: Implement simple helpers for construct type access.

### PR DESCRIPTION
**Pull request by PapyChacal**
Just provide `result_types` and `operands_types` on `Operation` and `args_types` on `Block` to get rid of a lot of (IMO) annoying comprehensions.

IRDLOperation hopefully following!

### CODE REVIEW

File: xdsl/dialects/gpu.py [hunk 1]
```diff
@@ -245,7 +245,7 @@ def verify_(self) -> None:
                 f"{self.operand.type}. They must be the same type for gpu.all_reduce"
             )
 
-        non_empty_body = any(b.ops for b in self.body.blocks)
+        non_empty_body = len(self.body.blocks) > 0
```
Comment (superlopuh) [inline xdsl/dialects/gpu.py:248]:
```suggestion
        non_empty_body = bool(self.body.blocks)
```

File: xdsl/ir/core.py [hunk 2]
```diff
@@ -614,6 +614,14 @@ class Operation(IRNode):
     def parent_node(self) -> IRNode | None:
         return self.parent
 
+    @property
+    def results_types(self) -> Sequence[Attribute]:
+        return tuple(r.type for r in self.results)
+
+    @property
+    def operands_types(self) -> Sequence[Attribute]:
+        return tuple(operand.type for operand in self.operands)
```
Comment (superlopuh) [inline xdsl/ir/core.py:623]: is this following a convention? I would expect `result_types` and `operand_types`, even though the original names are plurals. This follows `result_types` as the argument name to `build`, and `arg_types` in the Block init.

Comment (PapyChacal) [reply xdsl/ir/core.py:623]: Can do!

Comment (PapyChacal) [reply xdsl/ir/core.py]: Done!

**Pull request merged**